ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs

```diff
@@ -73,20 +81,121 @@ import qualified Shelley.Spec.Ledger.UTxO as SL
 
 import           Ouroboros.Consensus.Cardano.Block
 
+{-------------------------------------------------------------------------------
+  Figure out the transition point for Byron
+
+  The Byron ledger defines the update 'State' in
+  "Cardano.Chain.Update.Validation.Interface". The critical piece of state we
+  need is
+
+  > candidateProtocolUpdates :: ![CandidateProtocolUpdate]
+
+  which are the update proposals that have been voted on, accepted, and
+  endorsed, and now need to become stable. In `tryBumpVersion`
+  ("Cardano.Chain.Update.Validation.Interface.ProtocolVersionBump") we
+  find the candidates that are at least 'kUpdateStabilityParam' (@== 4k@) deep,
+  and then construct
+
+  > State
+  > { nextProtocolVersion    = cpuProtocolVersion
+  > , nextProtocolParameters = cpuProtocolParameters
+  > }
+
+  (with 'State' from "Cardano.Chain.Update.Validation.Interface.ProtocolVersionBump")
+  where 'cpuProtocolVersion'/'cpuProtocolParameters' are the version and
+  parameters from the update. This then ends up in the following callstack
+
+  > applyChainTick
+  > |
+  > \-- epochTransition
+  >     |
+  >     \-- registerEpoch
+  >         |
+  >         \-- tryBumpVersion
+
+  Now, if this is changing the major version of the protocol, then this actually
+  indicates the transition to Shelley, and the Byron 'applyChainTick' won't
+  actually happen. Instead, in 'singleEraTransition' we will report the
+  'EpochNo' of the transition as soon as it's @2k@ (not @4k@!) deep: in other
+  words, as soon as it is stable; at this point, the HFC will do the rest.
+
+  A slightly subtle point is that the Byron ledger does not record any
+  information about /past/ updates to the protocol parameters, and so if we
+  /were/ to ask the Byron ledger /after/ the update when the transition is
+  going to take place (did take place), it will say 'Nothing': transition not
+  yet known. In practice this won't matter, as it will have been translated to
+  a Shelley ledger at that point.
+-------------------------------------------------------------------------------}
+
+byronTransition :: PartialLedgerConfig ByronBlock
+                -> EraParams
+                -> Bound
+                -> LedgerState ByronBlock
+                -> Maybe EpochNo
+byronTransition ByronPartialLedgerConfig{..}
+                EraParams{..}
+                eraStart
+                st@ByronLedgerState{..} =
+      fmap cpuEpoch
+    . latest
+    . filter isStable
+    . CC.Update.candidateProtocolUpdates
+    . CC.cvsUpdateState
+    $ byronLedgerState
+  where
+    SecurityParam k = Byron.genesisSecurityParam byronLedgerConfig
+
+    cpuSlot :: CC.Update.CandidateProtocolUpdate -> SlotNo
+    cpuSlot = Byron.fromByronSlotNo . CC.Update.cpuSlot
+
+    -- This follows the same structure as the computation in the A/B test. Let
+    -- @s@ be the slot of the update proposal. Note that the very first slot in
```

Comment:
> Let @s@ be the slot of the update proposal.

That sounds like it's when the update was proposed. But it's when the update became a candidate (ie was endorsed).

I'm not sure it's worth rewording; I defer.

Reply:
Yeah, you're right.